src/test/java/com/znsio/reportPortal/RPReporter.java

```diff
@@ -0,0 +1,259 @@
+
+import java.util.*;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static com.google.common.base.Strings.isNullOrEmpty;
+
+class RPReporter {
+    private final Map<String, Date> featureStartDate = Collections.synchronizedMap(new HashMap<String, Date>());
+    private static final Logger LOGGER = Logger.getLogger(RPReporter.class.getName());
+    private Supplier<Launch> launch;
+
+    RPReporter() {
+    }
+
+    void startLaunch() {
+        this.launch = Suppliers.memoize(new Supplier<Launch>() {
+            private final Date startTime = getTime();
+
+            @Override
+            public Launch get() {
+                final ReportPortal reportPortal = ReportPortal.builder().build();
+                ListenerParameters parameters = reportPortal.getParameters();
+                StartLaunchRQ startLaunchRq = new StartLaunchRQ();
+                startLaunchRq.setName(parameters.getLaunchName());
+                startLaunchRq.setStartTime(startTime);
+                startLaunchRq.setMode(parameters.getLaunchRunningMode());
+                startLaunchRq.setAttributes(parameters.getAttributes());
+                if (!isNullOrEmpty(parameters.getDescription())) {
+                    startLaunchRq.setDescription(parameters.getDescription());
+                }
+                startLaunchRq.setRerun(parameters.isRerun());
+                if (!isNullOrEmpty(parameters.getRerunOf())) {
+                    startLaunchRq.setRerunOf(parameters.getRerunOf());
+                }
+                return reportPortal.newLaunch(startLaunchRq);
+            }
+        });
+        this.launch.get().start();
+    }
+
+    void finishLaunch(Suite suite) {
+        FinishExecutionRQ finishLaunchRq = new FinishExecutionRQ();
+        finishLaunchRq.setEndTime(getTime());
+        finishLaunchRq.setStatus(getLaunchStatus(suite));
+
+        launch.get().finish(finishLaunchRq);
+    }
+
+    synchronized boolean isTemplate(Feature feature) {
+        if (feature != null) {
+            List<Tag> featureTags = feature.getTags();
+            if (featureTags != null && !featureTags.isEmpty()) {
+                for (Tag tag : featureTags) {
+                    if (tag.getName().equalsIgnoreCase("template")) return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    synchronized boolean isTemplate(Scenario scenario) {
+        if (scenario != null) {
+            List<Tag> scenarioTags = scenario.getTags();
+            if (scenarioTags != null && !scenarioTags.isEmpty()) {
+                for (Tag tag : scenarioTags) {
+                    if (tag.getName().equalsIgnoreCase("template")) return true;
+                }
+            }
+        }
+        return false  ;
+    }
+
+    synchronized void startFeature(Feature feature) {
+        featureStartDate.put(getUri(feature), getTime());
+    }
+
+    synchronized void finishFeature(FeatureResult featureResult) {
```

Comment:
very complex method. break it up into smaller methods

Reply:
Method finishFeature has been broken down into smaller methods